politeia_sync.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil && err != storm.ErrNotFound {
+		return err
+	}
+
+	err = p.handleProposalsUpdate(proposals)
+	if err != nil {
+		return err
+	}
+
+	err = p.handleNewProposals(proposals)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Politeia) handleProposalsUpdate(proposals []Proposal) error {
+	if len(proposals) == 0 {
+		return nil
+	}
+
+	tokens := make([]string, len(proposals))
+	for i := range proposals {
+		tokens[i] = proposals[i].Token
+	}
+
+	for {
+		if len(tokens) == 0 {
+			break
+		}
+
+		var batch []string
+
+		limit := p.client.policy.ProposalListPageSize
+		if len(tokens) <= p.client.policy.ProposalListPageSize {
+			limit = len(tokens)
+		}
+
+		batch, tokens = tokens[:limit], tokens[limit:]
+
+		batchTokens := &Tokens{batch}
+		updatedProposals, err := p.client.batchProposals(batchTokens)
+		if err != nil {
+			return err
+		}
+
+		votesSummaries, err := p.client.batchVoteSummary(batchTokens)
+		if err != nil {
+			return err
+		}
+
+		for i := range updatedProposals {
+			if voteSummary, ok := votesSummaries.Summaries[updatedProposals[i].CensorshipRecord.Token]; ok {
+				updatedProposals[i].VoteSummary = voteSummary
+			}
+
+			for k := range proposals {
+				if proposals[k].CensorshipRecord.Token == updatedProposals[i].CensorshipRecord.Token {
+					if !reflect.DeepEqual(proposals[k], updatedProposals[i]) {
+						updatedProposals[i].ID = proposals[k].ID
+						err := p.db.Update(&updatedProposals)
+						if err != nil {
+							return err
+						}
+
+						if updatedProposals[i].VoteSummary.Status != proposals[k].VoteSummary.Status {
+							switch www.PropVoteStatusT(updatedProposals[i].VoteSummary.Status) {
+							case www.PropVoteStatusStarted:
+								p.onVoteStarted(&updatedProposals[i])
+							case www.PropVoteStatusFinished:
+								p.onVoteFinished(&updatedProposals[i])
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *Politeia) handleNewProposals(proposals []Proposal) error {
+	loadedTokens := make([]string, len(proposals))
+	for i := range proposals {
+		loadedTokens[i] = proposals[i].CensorshipRecord.Token
+	}
+
+	tokenInventory, err := p.client.tokenInventory()
+	if err != nil {
+		return err
+	}
+
+	return p.fetchAllUnfetchedProposals(tokenInventory, loadedTokens)
+}
+
+func (p *Politeia) onVoteStarted(proposal *Proposal) {
+	log.Info("Politeia sync: proposal vote status updated")
+	p.notificationListeners.OnVoteStarted(proposal.ID, proposal.CensorshipRecord.Token)
+}
+
+func (p *Politeia) onVoteFinished(proposal *Proposal) {
+	log.Info("Politeia sync: proposal vote status updated")
+	p.notificationListeners.OnVoteFinished(proposal.ID, proposal.CensorshipRecord.Token)
+}
+
+func (p *Politeia) onNewProposal(proposal *Proposal) {
+	log.Infof("Politeia sync: found new proposal %s", proposal.CensorshipRecord.Token)
+	p.notificationListeners.OnNewProposal(proposal.ID, proposal.CensorshipRecord.Token)
+}
+
+func diff(tokenInventory, savedTokens []string) []string {
```

Comment:
Make this function part of the `Politeia` struct and give it a name that better defines what it does. I initially thought it was a math function.